### ISSUE TITLE
commenting unused kernels can make code run

### ISSUE THREAD

**Issue by pepe78** [KhronosGroup/OpenCL-Docs#264]
So I am using OpenCL & Ubuntu [20.04 but I have seen it also happening on other versions as well] & NVIDIA GPU card. I have some kernels, which I compile successfully, but when running I get crash on copying memory from CPU to GPU. I comment out some unused kernels and then code runs just fine. This is rather weird... Why is code which is not used having influence if something can or cannot run?

**Comment by bashbaug** [comment]
Hi @pepe78, I've transferred your issue to OpenCL-Docs where it may receive more attention.

Some typical things to try when debugging a scenario like this:

1. Are any API calls returning errors?
2. When you're copying memory from the CPU to the GPU, are you copying the amount of memory you expect to be copying?

I usually use the [OpenCL Intercept Layer](https://github.com/intel/opencl-intercept-layer) to help answer these questions.

**Comment by pepe78** [comment]
Well, I said incorrectly "crash" - when copying CPU to GPU it does not return SUCCESS as it usually does. I am pretty sure I am copying correct amount of memory though.

**Comment by pepe78** [comment]
cl_int err = clEnqueueWriteBuffer(queue, gpu_mem, CL_TRUE, 0, size, cpu_mem, 0, NULL, NULL);
	if(err != CL_SUCCESS)

**Comment by ex-rzr** [comment]
1. What error do you get? (err != CL_SUCCESS is a bit vague)

2. Are you sure that the error happens in `clEnqueueWriteBuffer`? 
`clEnqueueWriteBuffer(..., CL_TRUE, ...)` is a blocking call, so it may return errors from any previous async calls. I suggest you to check `cl_int err = clFinish(queue)` before it.

**Comment by pepe78** [comment]
So I updated code:
void GPU::CopyCPUtoGPU(void *cpu_mem, cl_mem gpu_mem, int size)
{
	cl_int err = clFinish(queue);
	fprintf(stderr, "%d Current\n", ((int)err));
	err = clEnqueueWriteBuffer(queue, gpu_mem, CL_TRUE, 0, size, cpu_mem, 0, NULL, NULL);
	if(err != CL_SUCCESS)
	{
		fprintf(stderr, "%d %d %d %d %d %d %d %d\n", ((int)CL_SUCCESS), ((int)CL_INVALID_COMMAND_QUEUE), ((int)CL_INVALID_CONTEXT), ((int)CL_INVALID_MEM_OBJECT), ((int)CL_INVALID_VALUE), ((int)CL_INVALID_EVENT_WAIT_LIST), ((int)CL_MEM_OBJECT_ALLOCATION_FAILURE), ((int)CL_OUT_OF_HOST_MEMORY));
		fprintf(stderr, "%d Error copying cpu mem to gpu mem!\n", ((int)err));
		exit(-1);
	}
	WaitForGPUToFinish();
}

and receiving:
-36 Current
0 -36 -34 -38 -30 -57 -4 -6
-5 Error copying cpu mem to gpu mem!

not sure what -5 stands for (https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clEnqueueWriteBuffer.html), but -36 seems (from clFinish) to be CL_INVALID_COMMAND_QUEUE.

when I comment some unused code (kernels which are compiled but in this case are not executed), it runs just fine (without error).

Let me know if you would like steps to reproduce this weird behavior.

**Comment by pepe78** [comment]
After more debugging:
void GPU::WaitForGPUToFinish()
{
	cl_int err = clFinish(queue);
	if(err  != CL_SUCCESS)
	{
		fprintf(stderr, "Waiting for GPU failed %d!\n", err);
		exit(-1);
	}
}

Epoch 0 
Forward layer 0
Forward layer 1
Waiting for GPU failed -36!

CL_INVALID_COMMAND_QUEUE

I comment once again some unused kernels, and it runs just fine.

**Comment by ex-rzr** [comment]
If you want I can check how it runs on my 2060, but I need all steps to reproduce it, because my quick check with `./wots configBB.txt mnist/mnist_train.txt mnist/mnist_test.txt 4000 null 100 log none` seems to work fine.

**Comment by pepe78** [comment]
if you are already running wots then try this config:

convolution,1,28,28,10,5,5,0.5,-0.001
convolution,10,24,24,20,4,4,0.5,-0.001
max,200,21,21,7,7
matrix,1800,130,0.5,-0.001
sigmoid,130
dropout,130,0.5
matrix,130,10,0.5,-0.001
softmax,10

(two convolutions in row)

that will get you error.

now comment out functions augmentmatrix_forward and augmentmatrix_backward in  kernels.cl, and try it again. it will work...

**Comment by pepe78** [comment]
by the way - DeeperThought is exactly same as WideOpenThoughts but in CUDA C++. This weirdness happens only in OpenCL...

**Comment by ex-rzr** [comment]
Ok, I'm able to reproduce the bug.

And I guess, I know what's happening.

1. I compile the program
2. I run it but it fails with "Project needs to be recompiled with larger field for convolution layer" (line 19 in DNNLayerConvolution.cc)
3. So I increase `MAXX1X2` to 5760 in kernels.h and recompile
4. I run it again and it fails with OpenCL error
5. I comment those two kernels and it works fine

So what's happening:
NVidia implementation of OpenCL has one "feature" (this is not the first time I encounter this problem) - caching of compiled kernels, but with a HUGE FLAW: for detecting if the program requires recompilation it uses the program's source only and ignores any changes in included headers. 
(see https://stackoverflow.com/questions/31338520/opencl-clbuildprogram-caches-source-and-does-not-recompile-if-included-source)

After step 3 we have MAXX1X2 = 5760 in host code and 3920 in device code (because of the cached binary). When we comment those two kernels the source code changes, clBuildProgram really builds the program from sources, and everything works fine.
So this happens not because of those two kernels but any other change in kernels.cl (like a new line of a comment).

Workarounds this problem:
1. `rm -R ~/.nv/ComputeCache` everytime :)
2. run with `CUDA_CACHE_DISABLE=1 ./wots ...`
3. (**the best solution in my opinion**) remove `#include "./source/kernels.h"` and call `clCreateProgramWithSource` with contents of both source files (kernels.h and kernels.cl)

**Comment by pepe78** [comment]
wow - let me check

**Comment by pepe78** [comment]
so I did
1) moved kernels.h directly to kernels.cl
2) ran rm -R ~/.nv/ComputeCache
3) Now it fails on:
rm -R ~/.nv/ComputeCache

while I know about workarounds (and I have other workarounds), it seems still like this is an issue that should be fixed one day?

**Comment by pepe78** [comment]
from what I see - more kernels I add to kernels.cl - more weird issues is happening. Seems like OpenCL does not like predefined memory (float pics[AMMAXX1X2];).

Anyway - thanks for help.

**Comment by ex-rzr** [comment]
> Now it fails on:
Error enqueing kernel!

So you have the same values in `kernels.h` and `kernels.cl` but it still fails?

> Seems like OpenCL does not like predefined memory (float pics[AMMAXX1X2];).

Well, this should not be a problem if we're discussing correctness, but it's very likely that there is no performance gain to have `float pics[MAXX1X2]` (and perhaps `float convos[MAXNUMCONVY1Y2]`) because it is too large for storing in registers, so it is placed in global memory and I suspect that access to it has approximately the same performance as if you would read `inp` directly (but this needs profiling to be sure).

**Comment by pepe78** [comment]
well, it seemed to work for smaller memories (to speed up things), but i have not done profiling, so who knows.

but yeah - once i move kernels.h into kernels.cl, I see the enqueue error.

I guess my best work around is use DeeperThought, as I don't see these kind of issues when using CUDA C++. This is pity, as point to have WideOpenThoughts is able to run both NVIDIA and AMD GPUs. I will understand this as a problem of NVIDIA's implementation of OpenCL. Oh well.

**Comment by ex-rzr** [comment]
That's weird, I have the same changes as you (changed `#define MAXX1X2 5760`, rebuilt C++ and replaced `#include "./source/kernels.h"` with its content) and it works for me on GeForce 2060 (OpenCL 1.2 CUDA 10.2.159).
Here is my log: [log.txt](https://github.com/KhronosGroup/OpenCL-Docs/files/4532542/log.txt)

Are you sure that you've rebuilt the C++ code? (I've no other ideas why it happens on your side)

**Comment by pepe78** [comment]
oh shoot - my bad, inserting h file into cl file works. I had troubles as I was running on GPU also CUDA and it ran out of memory. Once I stopped that - it works just fine.

Seems like it really hates the h file. This is super weird. :) [this is better workaround though]

**Comment by pepe78** [comment]
another weird thing though - I go back and replace definitions with include, and it runs too... yikes

**Comment by bashbaug** [comment]
It seems like this issue is winding down.  Can we close it?  Thanks!

**Comment by pepe78** [comment]
it depends - issue is still there, but there are workarounds. I don't think issues should be closed until they are fixed. was something fixed on your side?

**Comment by bashbaug** [comment]
Thanks for the quick response!

If this issue is still occurring, is there a place it should move to instead?

This repo is for the OpenCL spec source and toolchain.  If there's something we should fix in the spec relating to this issue, I'm fine keeping it open here, but since this appears to be an issue in a specific implementation perhaps there's a better place to track it?

**Comment by pepe78** [comment]
I am not familiar with where is the best place to post this, as there are three possibilities where problem is:

1) using h file in .cl file
2) using large local array in kernel function
3) implementation of OpenCL on NVIDIA's card

it's hard to know which one it is (might be combination?). If you know where the best place is to report it, please feel free to move it there.

**Comment by pepe78** [comment]
As I have quite good workaround for this issue (do not use h files in cl files), I am closing this issue. Feel free to reopen it if you want, or give it low priority.
